### ISSUE TITLE
OnlineLearner implementation

### ISSUE THREAD

**Issue by sreckojoksimovic** [TeamCohen/MinorThird#1]
Hi,
I have read probably all the tutorials about classification, and I didn't have a problem to understand them. But, I need to use API for classification, since I already have an interface, and I cannot use MinorThird out of the box.

My messages are stored in database and I have 4 or 5 classes. The idea is to learn classifier using labeled messages, build model, and classify the rest of the messages.

I was thinking that OnlineLearner could be the appropriate solution, and I have looked at OnlineDemo.java. I would like to know if there are any disadvantages with this approach? It is not a problem to create a file for each message, but I would like to avoid that.

In the example, as I understood, are two classes (POS and NEG), what would be difference in my example (4 classes)?  

Best,
Srecko

**Comment by linfrank** [comment]
Online learner is only necessary if you want to update your learned model real-time (right after someone submits a new label via your UI). If your UI is only for classifying messages and not for learning new labels, then you should just train your models offline instead.

There are different online methods, but not all machine learning methods have an online version implemented, and offline methods allow for a tighter control over your training data (e.g., avoiding suffer the loss of accuracy due to bad input.)

You can turn any binary classifier into a multi-class classifier using something like OneVsAllClassifier - looking at the source for that class will give you a good idea what it does. If you are using the API and building your own system it's good to experiment on your own way of combining multiple binary classifiers into a multi-class one.

**Comment by sreckojoksimovic** [comment]
Thanks for the answer.

I think that this sentence: "the OnlineTextClassifierLearner allows you to add documents to a learner by passing in a document string rather than a document span" made me think of this approach as the better one.

I will take a more detailed look at OneVsAllClassifier as well as OneVsAllLearner, thanks for advice.

I guess that this means that I will have to store messages in files, and then use them for training and classification? I think 
it is not going to be such a problem, and I might be boring with that idea, but still want to be sure.

Once again, thank you for your time and tips.
Srecko

**Comment by linfrank** [comment]
if you use m3rd api for both training and classification then you can do whatever you want :)

**Comment by sreckojoksimovic** [comment]
I saw that just few hours ago... :) thanks
